### PR TITLE
[FLINK-34048][table] Support session window agg in table runtime instead of using legacy group window agg op

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

Currently, session window agg will rewrite to group window operator in exec node instead of using window operator. This pr is try to support session window tvf in table-runtime.

Note: This pr is a draft pr. The first and second commit is reviewing in other pr. 

## Brief change log

  - *add isAlignedWindow for WindowSpec interface*
  - *introduce basic interface and class for unslicing window*
  - *introduce unslicing window agg for session window tvf*
  - *add operator test for this pr*

## Verifying this change

1. There has been ITCases for session window tvf agg. So no need to add them. They are:
  - WindowAggregateITCase#testEventTimeSessionWindow
  - WindowAggregateITCase#testEventTimeSessionWindowWithCDCSource
  - WindowAggregateITCase#testDistinctAggWithMergeOnEventTimeSessionWindow
  - WindowAggregateTest#testSession_OnRowtime
  - WindowAggregateTest#testSession_OnProctime
  - WindowAggregateTest#testSession_DistinctSplitEnabled
  - WindowAggregateTest#testGroupKeyMoreThanPartitionKeyInSessionWindow
  - WindowAggregateTest#testGroupKeyLessThanPartitionKeyInSessionWindow
  - WindowAggregateJsonITCase#testEventTimeSessionWindow

3. Some harness tests for operator are added.
  - UnslicingWindowAggOperatorTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? An independent later pr will be introduced

### CODE REVIEW

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            } else {
+                aggregator.retract(element);
+            }
+            acc = aggregator.getAccumulators();
+            windowState.update(affectedWindow, acc);
+        }
+
+        // the actual window which the input row is belongs to
+        Optional<TimeWindow> actualWindowOp =
+                unsliceAssigner.assignActualWindow(element, clockService, windowFunction);
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:152]: Will the following case exist: affectedWindowOp is present, but actualWindowOp is empty? or the opposite?

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/aggregate/window/UnslicingWindowAggOperatorTest.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+                UnsliceAssigners.session(2, shiftTimeZone, Duration.ofSeconds(3));
+
+        final UnslicingSumAndCountAggsFunction aggsFunction =
+                new UnslicingSumAndCountAggsFunction();
+        UnslicingWindowOperator<RowData, ?> operator =
+                (UnslicingWindowOperator<RowData, ?>)
+                        WindowAggOperatorBuilder.builder()
+                                .inputSerializer(INPUT_ROW_SER)
+                                .shiftTimeZone(shiftTimeZone)
+                                .keySerializer(KEY_SER)
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/aggregate/window/UnslicingWindowAggOperatorTest.java:64]: Can you also test without key case?

**Review by lsyldliu — COMMENTED** [review]
@xuyangzhong Thanks for your contribution, I left some comments. Moreover, can you add some ITCase with partition key?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        }
+    }
+
+    @Override
+    public void clearWindow(long timerTimestamp, TimeWindow window) throws Exception {
+        windowFunction.cleanWindowIfNeeded(window, timerTimestamp);
+    }
+
+    @Override
+    public void advanceProgress(long progress) throws Exception {}
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:192]: Because this function is mainly used to flush buffered data into states, and currently, UnsliceWindowAggProcessor does not support Window Buffer yet.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnslicingTriggerContextBase.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+/**
+ * A base context for unslicing window trigger.
+ *
+ * <p>{@link UnslicingTriggerContextBase} is a utility for handling {@link Trigger} invocations. It
+ * can be reused by setting the {@code key} and {@code window} fields. No internal state must be
+ * kept in the {@link Trigger.TriggerContext}
+ */
+public abstract class UnslicingTriggerContextBase<K, W extends Window>
+        implements Trigger.TriggerContext, Trigger.OnMergeContext {
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnslicingTriggerContextBase.java:42]: You're right.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        super.open(context);
+        this.metrics = context.getRuntimeContext().getMetricGroup();
+        this.windowFunction =
+                new MergingWindowProcessFunction<>(
+                        unsliceAssigner.getInnerMergingWindowAssigner(),
+                        aggregator,
+                        unsliceAssigner
+                                .getInnerMergingWindowAssigner()
+                                .getWindowSerializer(new ExecutionConfig()),
+                        0L);
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:107]: Because all window tvf do not support experimental options such as early fire, late fire and allow lateness.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java (hunk 1 above)
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:152]: That is impossible. I have added a check `Preconditions.checkArgument` here.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnslicingWindowProcessor.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package org.apache.flink.table.runtime.operators.window.tvf.unslicing;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.runtime.operators.window.tvf.common.WindowProcessor;
+
+/** A processor that processes elements for unslicing windows. */
+@Internal
+public interface UnslicingWindowProcessor<W> extends WindowProcessor<W> {}
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnslicingWindowProcessor.java:26]: Because we can't let developer to create a class only implement `UnslicingWindowProcessor` but not implement `WindowProcessor` .

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+/**
+ * An window aggregate processor implementation which works for {@link UnsliceAssigner}, e.g.
+ * session windows.
+ */
+public class UnsliceWindowAggProcessor extends AbstractWindowAggProcessor<TimeWindow>
+        implements UnslicingWindowProcessor<TimeWindow> {
+
+    private final UnsliceAssigner<TimeWindow> unsliceAssigner;
+
+    private final Trigger<TimeWindow> trigger;
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:65]: Retain it after removing AbstractUnslicingTriggerContextBase.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnslicingTriggerContextBase.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.Collection;
+
+/**
+ * A base context for unslicing window trigger.
+ *
+ * <p>{@link UnslicingTriggerContextBase} is a utility for handling {@link Trigger} invocations. It
+ * can be reused by setting the {@code key} and {@code window} fields. No internal state must be
+ * kept in the {@link Trigger.TriggerContext}
+ */
+public abstract class UnslicingTriggerContextBase<K, W extends Window>
```
[the remark this answers is not in the document]
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnslicingTriggerContextBase.java:41]: I'll remove this abstract class.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnsliceAssigners.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                return Optional.empty();
+            }
+        }
+
+        protected long getUtcTimestamp(RowData element, ClockService clock) {
+            final long timestamp;
+            if (rowtimeIndex >= 0) {
+                if (element.isNullAt(rowtimeIndex)) {
+                    throw new RuntimeException(
+                            "RowTime field should not be null,"
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/tvf/unslicing/UnsliceAssigners.java:124]: It would be better also throw the `rowtimeIndex`.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+            triggerContext.onMerge();
+        }
+
+        @Override
+        public <S extends State> S getPartitionedState(StateDescriptor<S, ?> stateDescriptor)
+                throws Exception {
+            requireNonNull(stateDescriptor, "The state properties must not be null");
+            return ctx.getKeyedStateBackend()
+                    .getPartitionedState(
+                            VoidNamespace.INSTANCE,
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:271]: Why the namespace is void?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/AbstractWindowAggProcessor.java [hunk 11]
```diff
@@ -74,6 +74,9 @@ public abstract class AbstractWindowAggProcessor<W> implements WindowProcessor<W
     /** state schema: [key, window, accumulator]. */
     protected transient WindowValueState<W> windowState;
 
+    /** The inner state in {@link #windowState}. */
+    protected transient InternalValueState<RowData, W, RowData> internalWindowState;
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/AbstractWindowAggProcessor.java:78]: No one use this member variable, I think we can revert this change.

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/aggregate/window/UnslicingWindowAggOperatorTest.java [hunk 12]
```diff
@@ -0,0 +1,628 @@
...
+        ASSERTER.assertOutputEqualsSorted(
+                "Output was not correct.", expectedOutput, testHarness.getOutput());
+
+        assertThat(operator.getNumLateRecordsDropped().getCount()).isEqualTo(1);
+
+        testHarness.close();
+    }
+
+    @Test
+    public void testEventTimeSessionWindowsWithCDC() throws Exception {
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/aggregate/window/UnslicingWindowAggOperatorTest.java:172]: CDC -> Changelog?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGlobalWindowAggregate.java [hunk 13]
```diff
@@ -169,7 +171,11 @@ protected Transformation<RowData> translateToPlanInternal(
                 TimeWindowUtil.getShiftTimeZone(
                         windowing.getTimeAttributeType(),
                         TableConfigUtils.getLocalTimeZone(config));
-        final SliceAssigner sliceAssigner = createSliceAssigner(windowing, shiftTimeZone);
+        final WindowAssigner windowAssigner = createWindowAssigner(windowing, shiftTimeZone);
```
Comment (lsyldliu) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGlobalWindowAggregate.java:174]: What is the behavior if enable the two-phase agg forcibly with session window? Assuming the following pattern:
Local Agg -> Global Agg -> Session Window? Does it can work?

**Review by lsyldliu — COMMENTED** [review]
@xuyangzhong Thanks for update, I left some comments.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java (hunk 10 above)
Comment (xuyangzhong) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java:271]: If the namespace is `Void`, that means the state gotten should not care about the namespace(window). In this case, this code is used for get state for `mapping` in `MergingWindowSet`. It does not pay attention to what the specific window is.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGlobalWindowAggregate.java (hunk 13 above)
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecGlobalWindowAggregate.java:174]: Session window agg does not support two-stage optimization yet.

**Review by lsyldliu — APPROVED** [review]
LGTM
